File: backend/src/utils/email.rs
use lettre::message::header::ContentType;
use lettre::transport::smtp::authentication::Credentials;
use lettre::{Message, SmtpTransport, Transport};
use rand::Rng;
use std::error::Error;
use tracing::{error, info};

use crate::core::configs::AppConfig;

/// Generate a 6-digit OTP code
pub fn generate_otp() -> String {
    let mut rng = rand::thread_rng();
    let otp: u32 = rng.gen_range(100000..999999);
    otp.to_string()
}

/// Send welcome email with OTP
///
/// # Arguments
/// * `config` - Application configuration containing SMTP settings
/// * `to_email` - Recipient email address
/// * `name` - Recipient's name
/// * `otp` - The OTP code to include in the email
///
/// # Returns
/// Result indicating success or failure
pub async fn send_welcome_email(
    config: &AppConfig,
    to_email: &str,
    name: &str,
    otp: &str,
) -> Result<(), Box<dyn Error + Send + Sync>> {
    // Read the welcome.html template
    let template = include_str!("../templates/welcome.html");
    
    // Replace placeholders - only name and OTP code
    let html_body = template
        .replace("[Name]", name)
        .replace("[OTP_CODE]", otp);

    // Build the email
    let email = Message::builder()
        .from(
            config
                .smtp_from_email
                .parse()
                .map_err(|e| format!("Invalid from email: {}", e))?,
        )
        .to(to_email
            .parse()
            .map_err(|e| format!("Invalid to email: {}", e))?)
        .subject("Welcome to Here - Verify Your Account")
        .header(ContentType::TEXT_HTML)
        .body(html_body)?;

    // Create SMTP credentials
    let creds = Credentials::new(
        config.smtp_username.clone(),
        config.smtp_password.clone(),
    );

    // Build the SMTP transport
    let mailer = SmtpTransport::relay(&config.smtp_host)?
        .port(config.smtp_port)
        .credentials(creds)
        .build();

    // Send the email
    match mailer.send(&email) {
        Ok(_) => {
            info!("Welcome email sent successfully to {}", to_email);
            Ok(())
        }
        Err(e) => {
            error!("Failed to send welcome email to {}: {}", to_email, e);
            Err(Box::new(e))
        }
    }
}

/// Send welcome email for verified users (OAuth signups)
///
/// # Arguments
/// * `config` - Application configuration containing SMTP settings
/// * `to_email` - Recipient email address
/// * `name` - Recipient's name
///
/// # Returns
/// Result indicating success or failure
pub async fn send_verified_welcome_email(
    config: &AppConfig,
    to_email: &str,
    name: &str,
) -> Result<(), Box<dyn Error + Send + Sync>> {
    // Read the welcome-verified.html template
    let template = include_str!("../templates/welcome-verified.html");
    
    // Replace placeholders - only name
    let html_body = template.replace("[Name]", name);

    // Build the email
    let email = Message::builder()
        .from(
            config
                .smtp_from_email
                .parse()
                .map_err(|e| format!("Invalid from email: {}", e))?,
        )
        .to(to_email
            .parse()
            .map_err(|e| format!("Invalid to email: {}", e))?)
        .subject("Welcome to H.E.R.E - Let's Get Started!")
        .header(ContentType::TEXT_HTML)
        .body(html_body)?;

    // Create SMTP credentials
    let creds = Credentials::new(
        config.smtp_username.clone(),
        config.smtp_password.clone(),
    );

    // Build the SMTP transport
    let mailer = SmtpTransport::relay(&config.smtp_host)?
        .port(config.smtp_port)
        .credentials(creds)
        .build();

    // Send the email
    match mailer.send(&email) {
        Ok(_) => {
            info!("Verified welcome email sent successfully to {}", to_email);
            Ok(())
        }
        Err(e) => {
            error!("Failed to send verified welcome email to {}: {}", to_email, e);
            Err(Box::new(e))
        }
    }
}

/// Send OTP verification email
///
/// # Arguments
/// * `config` - Application configuration containing SMTP settings
/// * `to_email` - Recipient email address
/// * `otp` - The OTP code to send
///
/// # Returns
/// Result indicating success or failure
pub async fn send_otp_email(
    config: &AppConfig,
    to_email: &str,
    otp: &str,
) -> Result<(), Box<dyn Error + Send + Sync>> {
    // Read the otp.html template
    let template = include_str!("../templates/otp.html");
    
    // Replace placeholders - only OTP code
    let html_body = template.replace("[OTP_CODE]", otp);

    // Build the email
    let email = Message::builder()
        .from(
            config
                .smtp_from_email
                .parse()
                .map_err(|e| format!("Invalid from email: {}", e))?,
        )
        .to(to_email
            .parse()
            .map_err(|e| format!("Invalid to email: {}", e))?)
        .subject("Your Here Verification Code")
        .header(ContentType::TEXT_HTML)
        .body(html_body)?;

    // Create SMTP credentials
    let creds = Credentials::new(
        config.smtp_username.clone(),
        config.smtp_password.clone(),
    );

    // Build the SMTP transport
    let mailer = SmtpTransport::relay(&config.smtp_host)?
        .port(config.smtp_port)
        .credentials(creds)
        .build();

    // Send the email
    match mailer.send(&email) {
        Ok(_) => {
            info!("OTP email sent successfully to {}", to_email);
            Ok(())
        }
        Err(e) => {
            error!("Failed to send OTP email to {}: {}", to_email, e);
            Err(Box::new(e))
        }
    }
}

/// Store OTP in Redis with expiration
///
/// # Arguments
/// * `redis_pool` - Redis connection pool
/// * `prefix` - Redis key prefix (e.g., "otp:")
/// * `email` - User's email address (used as part of the key)
/// * `otp` - The OTP code to store
/// * `expiry_seconds` - TTL in seconds
///
/// # Returns
/// Result indicating success or failure
pub async fn store_otp_in_redis(
    redis_pool: &deadpool_redis::Pool,
    prefix: &str,
    email: &str,
    otp: &str,
    expiry_seconds: u64,
) -> Result<(), Box<dyn Error + Send + Sync>> {
    use deadpool_redis::redis::AsyncCommands;

    let mut conn = redis_pool.get().await?;
    let key = format!("{}{}", prefix, email);
    
    // Set the OTP with expiration
    let _: () = conn.set_ex(&key, otp, expiry_seconds as u64).await?;
    
    info!("OTP stored in Redis for {} with {}s expiry", email, expiry_seconds);
    Ok(())
}

/// Verify OTP from Redis
///
/// # Arguments
/// * `redis_pool` - Redis connection pool
/// * `prefix` - Redis key prefix (e.g., "otp:")
/// * `email` - User's email address
/// * `otp` - The OTP code to verify
///
/// # Returns
/// Result with bool indicating if OTP is valid
pub async fn verify_otp_from_redis(
    redis_pool: &deadpool_redis::Pool,
    prefix: &str,
    email: &str,
    otp: &str,
) -> Result<bool, Box<dyn Error + Send + Sync>> {
    use deadpool_redis::redis::AsyncCommands;

    let mut conn = redis_pool.get().await?;
    let key = format!("{}{}", prefix, email);
    
    // Get the stored OTP
    let stored_otp: Option<String> = conn.get(&key).await?;
    
    match stored_otp {
        Some(stored) if stored == otp => {
            // OTP is valid, delete it so it can't be reused
            let _: () = conn.del(&key).await?;
            info!("OTP verified and deleted for {}", email);
            Ok(true)
        }
        Some(_) => {
            info!("Invalid OTP provided for {}", email);
            Ok(false)
        }
        None => {
            info!("No OTP found or expired for {}", email);
            Ok(false)
        }
    }
}

/// Blacklist a token in Redis with expiration
///
/// # Arguments
/// * `redis_pool` - Redis connection pool
/// * `prefix` - Redis key prefix (e.g., "blacklist:")
/// * `token` - The token to blacklist
/// * `expiry_seconds` - TTL in seconds (should match token expiry)
///
/// # Returns
/// Result indicating success or failure
pub async fn blacklist_token(
    redis_pool: &deadpool_redis::Pool,
    prefix: &str,
    token: &str,
    expiry_seconds: u64,
) -> Result<(), Box<dyn Error + Send + Sync>> {
    use deadpool_redis::redis::AsyncCommands;

    let mut conn = redis_pool.get().await?;
    let key = format!("{}{}", prefix, token);
    
    // Store a marker value with expiration
    let _: () = conn.set_ex(&key, "1", expiry_seconds).await?;
    
    info!("Token blacklisted with {}s expiry", expiry_seconds);
    Ok(())
}

/// Check if a token is blacklisted
///
/// # Arguments
/// * `redis_pool` - Redis connection pool
/// * `prefix` - Redis key prefix (e.g., "blacklist:")
/// * `token` - The token to check
///
/// # Returns
/// Result with bool indicating if token is blacklisted
pub async fn is_token_blacklisted(
    redis_pool: &deadpool_redis::Pool,
    prefix: &str,
    token: &str,
) -> Result<bool, Box<dyn Error + Send + Sync>> {
    use deadpool_redis::redis::AsyncCommands;

    let mut conn = redis_pool.get().await?;
    let key = format!("{}{}", prefix, token);
    
    // Check if the key exists
    let exists: bool = conn.exists(&key).await?;
    
    Ok(exists)
}

File: backend/app/utils/email.py
import email.mime.text
import logging
import pathlib
import random
from typing import Optional
import aiosmtplib
import redis.asyncio as aioredis
from app.config import settings

logger = logging.getLogger("app.email")


# --- 1. Helper utilities ---
def generate_otp() -> str:
    """Generate a cryptographically secure 6-digit OTP code."""
    return str(random.randint(100000, 999999))


# --- 2. HTML template renderer ---
def render_template(template_name: str, replacements: dict) -> str:
    """Render an HTML email template from the templates directory."""
    # Build absolute path to template
    template_path = (
        pathlib.Path(__file__).parent.parent / "templates" / template_name
    )

    if not template_path.exists():
        # Fallback if templates haven't been copied yet
        fallback_templates = {
            "welcome.html": "<html><body><h1>Welcome!</h1><p>Hi [Name], your OTP code is: <strong>[OTP_CODE]</strong></p></body></html>",
            "welcome-verified.html": "<html><body><h1>Welcome!</h1><p>Hi [Name], your account has been successfully verified.</p></body></html>",
            "otp.html": "<html><body><h1>Verification Code</h1><p>Your verification code is: <strong>[OTP_CODE]</strong></p></body></html>",
        }
        html = fallback_templates.get(template_name, "<html><body></body></html>")
    else:
        with open(template_path, "r", encoding="utf-8") as f:
            html = f.read()

    # Replace placeholders
    for key, value in replacements.items():
        html = html.replace(f"[{key}]", str(value))
        html = html.replace(key, str(value))  # direct matching

    return html


# --- 3. Async SMTP and Logging Mailer ---
async def send_async_email(
    to_email: str, subject: str, html_body: str
) -> None:
    """Send an HTML email asynchronously using SMTP, with a console log fallback."""
    # Determine if dummy/local config is active
    is_mock = (
        settings.smtp_host == "localhost"
        or settings.smtp_username == "test_user"
        or settings.smtp_username == "user@gmail.com"
        or settings.smtp_password == "app-specific-password"
        or not settings.smtp_host
    )

    if is_mock:
        logger.info(
            f"\n"
            f"========================================================================\n"
            f"                       [DEV MAIL NOTIFICATION]\n"
            f"To: {to_email}\n"
            f"Subject: {subject}\n"
            f"Body:\n"
            f"{html_body}\n"
            f"========================================================================"
        )
        return

    # Build standard MIME email message
    msg = email.mime.text.MIMEText(html_body, "html")
    msg["Subject"] = subject
    msg["From"] = settings.smtp_from_email
    msg["To"] = to_email

    try:
        # Connect and send via async SMTP using aiosmtplib.
        # Passing username and password directly allows aiosmtplib to handle
        # connection, TLS/STARTTLS, and authentication automatically and robustly.
        async with aiosmtplib.SMTP(
            hostname=settings.smtp_host,
            port=settings.smtp_port,
            username=settings.smtp_username or None,
            password=settings.smtp_password or None,
            use_tls=(settings.smtp_port == 465),
        ) as smtp:
            await smtp.send_message(msg)
        logger.info(f"Email with subject '{subject}' sent successfully to {to_email}")
    except Exception as e:
        logger.error(f"Failed to send email to {to_email} via SMTP: {e}")
        # Log to console as emergency fallback
        logger.warning(
            f"[FALLBACK MAIL LOG]\nTo: {to_email}\nSubject: {subject}\nBody: {html_body}"
        )


async def send_welcome_email(to_email: str, name: str, otp: str) -> None:
    """Send a welcome email with OTP code."""
    html = render_template(
        "welcome.html", {"[Name]": name, "[OTP_CODE]": otp, "OTP_CODE": otp}
    )
    await send_async_email(to_email, "Welcome to Here - Verify Your Account", html)


async def send_verified_welcome_email(to_email: str, name: str) -> None:
    """Send welcome email to pre-verified third-party signups."""
    html = render_template("welcome-verified.html", {"[Name]": name})
    await send_async_email(
        to_email, "Welcome to H.E.R.E - Let's Get Started!", html
    )


async def send_otp_email(to_email: str, otp: str) -> None:
    """Send an OTP code verification email."""
    html = render_template("otp.html", {"[OTP_CODE]": otp, "OTP_CODE": otp})
    await send_async_email(to_email, "Your Here Verification Code", html)


# --- 4. Redis Async Integrations ---
async def store_otp_in_redis(
    redis_client: aioredis.Redis, email: str, otp: str, expiry_seconds: int
) -> None:
    """Store verification OTP code in Redis with TTL expiration."""
    key = f"{settings.otp_prefix}{email}"
    await redis_client.setex(key, expiry_seconds, otp)
    logger.info(
        f"OTP stored in Redis for {email} with {expiry_seconds}s expiry"
    )


async def verify_otp_from_redis(
    redis_client: aioredis.Redis, email: str, otp: str
) -> bool:
    """Verify OTP code from Redis. Deletes it if verification succeeds."""
    key = f"{settings.otp_prefix}{email}"
    stored_otp = await redis_client.get(key)
    if stored_otp is not None:
        stored_otp = stored_otp if isinstance(stored_otp, str) else stored_otp.decode("utf-8")

    if stored_otp == otp:
        await redis_client.delete(key)
        logger.info(f"OTP verified and deleted for {email}")
        return True

    logger.info(f"OTP verification failed for {email}")
    return False


async def blacklist_token(
    redis_client: aioredis.Redis, token: str, expiry_seconds: int
) -> None:
    """Blacklist a JWT token in Redis with a TTL matching remaining token lifespan."""
    key = f"{settings.blacklist_token_prefix}{token}"
    await redis_client.setex(key, expiry_seconds, "1")
    logger.info(f"Token blacklisted with {expiry_seconds}s expiry")


async def is_token_blacklisted(
    redis_client: aioredis.Redis, token: str
) -> bool:
    """Check if token is blacklisted in Redis."""
    key = f"{settings.blacklist_token_prefix}{token}"
    exists = await redis_client.exists(key)
    return bool(exists)
